Implement replay analysis support during spectating

Sheppsu opened the pull request:
Implements the replay analysis overlay into both solo and multiplayer spectating. For multiplayer spectating, the settings container is added on the left side under the leaderboard. For solo spectating it works the same as watching a replay.

Look of the settings container during multiplayer spectating:
![II4uig8bm6o](https://github.com/user-attachments/assets/243e36ac-9de9-4da7-a77f-5643e24624a3)

Sheppsu committed:
implement replay analysis into spectating

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs:
```diff
@@ -73,10 +78,24 @@ public MultiSpectatorScreen(Room room, MultiplayerRoomUser[] users)
+        /// Add a settings group to the bottom of the side container. Intended to be used by rulesets to add spectate-specific settings.
+        /// </summary>
+        /// <param name="settings">The settings group to be shown.</param>
+        public void AddSettingsAsync(PlayerSettingsGroup settings)
```
Why is this async?

Sheppsu replied:
When I originally tried adding the settings container from load there were problems with each `DrawableOsuRuleset` adding a container, even checking `SettingsAdded` in `MultiSpectatorScreen`, so I figured it needed to be added from `LoadComplete` and that I should also use `LoadComponenetAsync` in that case (maybe wrong); however, I can't call it from `DrawableOsuRuleset` because it needs to first verify via `SettingsAdded` that a settings container wasn't already added.

bdach replied:
> there were problems with each `DrawableOsuRuleset` adding a container

What problems specifically?

If those can be solved maybe this can happen synchronously and a whole bunch of bool-tracking badness could go away in favour of just a simple check.

Sheppsu replied:
Actually, thinking about it again, I think I've suddenly realized why that was occurring... I'll try to put it back in load with synchronous loading

Sheppsu replied:
My idea did not fix it... currently it looks like this in `DrawableOsuRuleset`:
```CSharp
if (multiSpectatorScreen == null)
    player.AddSettings(new ReplayAnalysisSettings(Config));
else if (!multiSpectatorScreen.SettingsAdded)
    multiSpectatorScreen.AddSettings(new ReplayAnalysisSettings(Config));
```
and `AddSettings` is basically the same, just using `Schedule` and no `LoadComponentAsync`

The problem is that multiple analysis settings get added to the screen.

Sheppsu replied:
Using a lock fixes the problem

bdach commented on osu.Game.Rulesets.Osu/UI/ReplayAnalysisOverlay.cs:
```diff
@@ -27,6 +27,8 @@ public partial class ReplayAnalysisOverlay : CompositeDrawable
 
         private readonly Replay replay;
 
+        private int replayFrameIndex = 0;
```
What is this for?

Sheppsu replied:
It's to keep track of the last frame that was processed for adding entries since during spectating the full replay isn't there

bdach replied:
The whole structure needs a reshuffle, you can't just tack on stuff like that. `initialise()` is not actually an initialisation method (it borderline wasn't before, what with it being called in `Update()`, but it *was* still mostly initialisation logic - now this method is executing *persistent* logic based on incoming frames). Yes you split the replay processing to `addEntries()` but the only caller of the method is....... `initialise()`.

Sheppsu replied:
makes sense

bdach commented on osu.Game.Rulesets.Osu/UI/ReplayAnalysisOverlay.cs:
```diff
@@ -75,7 +83,12 @@ protected override void Update()
         {
             if (loaded.IsValid)
+            {
+                if (CursorPath != null)
```
What is this conditional doing?

Sheppsu replied:
If the analysis containers are loaded, check for new frames to add entries for

bdach replied:
...why is that being achieved via checking `CursorPath`? without any inline commentary? this is so opaque.

whatever this is doing, it needs to be done better. preferably in a way that does not require explanation. even with your explanation above i can barely make out what that is supposed to be doing.

Sheppsu replied:
Sorry, I should've clarified better. It was to make sure the container variables weren't null (just checking one would be enough), though I just realized if loaded is valid then the containers are also not null.

bdach reviewed (commented):
i have several questions

Sheppsu committed:
fix structure of update functions

Sheppsu committed:
remove unnecessary invalidateLoaded call

Sheppsu committed:
remove unnecessary token register

Sheppsu committed:
load analysis settings synchronously

specifically in multiplayer spectating

Sheppsu committed:
slight restructure + comments

Sheppsu commented:
Not sure if parsing new frames during each update instead of all at the beginning means the logic in addEntries should be changed to async now, though I haven't had any performance problems running it.

Also now that it's not all parsed at once, would it be better to limit how far ahead it adds entries for? The purpose being to lessen the initial strain for larger replays being viewed in solo.

peppy commented:
Please don't do that until it's necessary.

Sheppsu commented:
> Please don't do that until it's necessary.

Didn't plan to; that's why I'm commenting about it.

bdach commented on osu.Game.Rulesets.Osu/UI/DrawableOsuRuleset.cs:
```diff
@@ -40,12 +41,17 @@ public DrawableOsuRuleset(Ruleset ruleset, IBeatmap beatmap, IReadOnlyList<Mod>?
-                replayPlayer.AddSettings(new ReplayAnalysisSettings(Config));
+                PlayfieldAdjustmentContainer.Add(new ReplayAnalysisOverlay(player.Score.Replay));
+
+                // if in multiplayer spectator mode, place the settings elsewhere
+                if (multiSpectatorScreen == null)
+                    player.AddSettings(new ReplayAnalysisSettings(Config));
+                else if (!multiSpectatorScreen.SettingsAdded)
+                    multiSpectatorScreen.AddSettings(() => new ReplayAnalysisSettings(Config));
```
I was *sort of* fine with this when it only affected replay player but this I am not fine with. It's not sane.

If the replay analysis settings are to be displayed on the multiplayer spectator screen then they should be added there _by the multiplayer spectator screen itself_. Probably via some virtual method on `Ruleset` that normally does nothing.

Sheppsu replied:
I agree with this, and in that case I think it'd make sense to have a non-ruleset-specific `AnalysisSettings`, which rulesets can create their own from and override a virtual method for creating the settings. If that is done, I think the `ReplayPlayer` should also add settings itself. Although, would it not make more sense to have the virtual method in `DrawableRuleset` as opposed to `Ruleset`?

bdach replied:
> Although, would it not make more sense to have the virtual method in `DrawableRuleset` as opposed to `Ruleset`?

We haven't done that sort of virtual in `DrawableRuleset` thus far. It would definitely need to be checked that constructing the drawable ruleset first to get the controls doesn't cause some massive overheads.

Sheppsu replied:
Yeah that could cause a problem. On second thought anyway, I guess `DrawableRuleset` is more for gameplay drawing(?), while the analysis settings are separate from gameplay, so putting it in `Ruleset` seems fine.

bdach commented on osu.Game.Rulesets.Osu/UI/ReplayAnalysisOverlay.cs:
```diff
@@ -90,14 +111,23 @@ private void initialise()
                 FrameMarkers = new FrameMarkerContainer(),
             };
 
+            LoadComponentsAsync(newDrawables, drawables =>
+            {
+                InternalChildrenEnumerable = drawables;
+                loadState.Value = true;
+            }, generationCancellationSource.Token);
```
I don't get any of this still. Why is this async-loading the content now? What is that needed for?

Sheppsu replied:
The drawables were always async-loaded ([master](https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Osu/UI/ReplayAnalysisOverlay.cs#L125)). Peppy did that part, but I _assume_ it's because the drawables are reloaded every time the display length changes, so multiple changes in a short time-span would cause lag. 
I'm not familiar enough with the core logic behind lifetimes, but I am wondering, could `LifetimeEnd` in `AnalysisFrameEntry` be changed during replay (probably by a bindable), instead of reloading the entire overlay each time? I think that way the async-loading wouldn't be necessary.

Sheppsu committed:
move replay analysis settings creation point

replay analysis settings is created in the respective drawable that's adding it, instead of adding it from `DrawableRuleset`
for this, it adds a virtual method in Ruleset that uses a non-ruleset-specific `ReplayAnalysisSettings` (the osu specific one was renamed to `OsuReplayAnalysisSettings`).

Sheppsu committed:
Merge branch 'master' into replay-analysis-spectating

# Conflicts:
#	osu.Game.Rulesets.Osu/UI/DrawableOsuRuleset.cs

Sheppsu commented:
oops, messed up trying to fix the merge conflict 😓

Sheppsu committed:
Revert "Merge branch 'master' into replay-analysis-spectating"

This reverts commit c5f491a3944b08da173fc7d344dee0ee2f13e3ee.

Sheppsu committed:
Merge branch 'master' into replay-analysis-spectating

Sheppsu committed:
Merge branch 'master' into replay-analysis-spectating

Theighlin commented:
The fact that you can change replay analysis from this screen but not visual settings (background dim, beatmap skin...) feels weird, the latter is probably more important than the former (it's diffuclt to see the benefits of analysis in realtime). I think a pop up overlay from the bottom or the right would work better, kinda like stable, if that's not seen as bad UX in lazer (players might want it in solo play too)

peppy commented:
Yeah that placement of the control on the left is not going to float.

Sheppsu commented:
That's valid, I just didn't come up with a better place to put it

peppy commented:
I'd say we want to use the standard placement as found in spectator view. The settings can be minimised by default, or they can slide in from the right if even when minimised they are too distracting.

Sheppsu commented:
By standard placement do you mean inside the players? or to the right of the player grid?

peppy commented:
> By standard placement do you mean inside the players? or to the right of the player grid?

Just overlapping the player grid basically. And hopefully not too in-the-way due to being collapsed.

Sheppsu committed:
Merge branch 'master' into replay-analysis-spectating

Sheppsu committed:
add settings to player settings overlay

Sheppsu committed:
rely on virtual method instead of player type

Sheppsu committed:
rename and document virtual method
